test/e2e.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	return tmpfile.Name(), nil
+}
+
+func runPromotion(cwd string, manifestPath string) error {
+	defer os.Remove(manifestPath)
+
+	// NOTE: we should probably compile the cip binary and then put it in the
+	// PATH and then use multirun.sh as-is.
+
+	cmd := exec.Command(
+		"bazel",
+		"run",
+		"--workspace_status_command="+cwd+"/workspace_status.sh",
+		":cip",
+		"--",
+		"-dry-run=false",
+		"-verbosity=3",
+		"-manifest="+manifestPath)
+
+	cmd.Dir = cwd
+
+	var stdout bytes.Buffer
+	cmd.Stdout = &stdout
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	if err != nil {
+		fmt.Println("stdout", stdout.String())
+		fmt.Println("stderr", stderr.String())
+		return err
+	}
+
+	fmt.Println(stdout.String())
+	return nil
+}
+
+func extractSvcAcc(registry reg.RegistryName, mfest reg.Manifest) string {
+	for _, r := range mfest.Registries {
+		if r.Name == registry {
+			return r.ServiceAccount
+		}
+	}
+	return ""
+}
+
+func getSnapshot(cwd string,
+	registry reg.RegistryName,
+	mfest reg.Manifest) ([]reg.Image, error) {
+
+	invocation := []string{
+		"run",
+		"--workspace_status_command=" + cwd + "/workspace_status.sh",
+		":cip",
+		"--",
+		"-snapshot=" + string(registry)}
+
+	svcAcc := extractSvcAcc(registry, mfest)
+	if len(svcAcc) > 0 {
+		invocation = append(invocation, "-snapshot-service-account="+svcAcc)
+	}
+
+	cmd := exec.Command("bazel", invocation...)
+
+	cmd.Dir = cwd
+
+	var stdout bytes.Buffer
+	cmd.Stdout = &stdout
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	if err != nil {
+		fmt.Println("stdout", stdout.String())
+		fmt.Println("stderr", stderr.String())
+		return nil, err
+	}
+
+	var images []reg.Image
+	if err = yaml.UnmarshalStrict(stdout.Bytes(), &images); err != nil {
+		return nil, err
+	}
+	return images, err
+}
+
+func clearRepository(regName reg.RegistryName,
+	mfest reg.Manifest,
+	sc *reg.SyncContext) {
+
+	mkDeletionCmd := func(
+		dest reg.RegistryContext,
+		imageName reg.ImageName,
+		digest reg.Digest) stream.Producer {
+		var sp stream.Subprocess
+		sp.CmdInvocation = reg.GetDeleteCmd(
+			dest,
+			sc.UseServiceAccount,
+			imageName,
+			digest,
+			true)
+		return &sp
+	}
+
+	sc.ClearRepository(regName, mfest, mkDeletionCmd, nil)
+}
+
+// E2ETest holds all the information about a single e2e test. It has the
+// promoter manifest, and the before/after snapshots of all repositories that it
+// cares about.
+type E2ETest struct {
+	Name      string             `name:"tests,omitempty"`
+	Manifest  reg.Manifest       `manifest:"tests,omitempty"`
+	Snapshots []RegistrySnapshot `snapshots:"tests,omitempty"`
+}
+
+// E2ETests is an array of E2ETest.
+type E2ETests []E2ETest
+
+// RegistrySnapshot is the snapshot of a registry. It is basically the key/value
+// pair out of the reg.MasterInventory type (RegistryName + []Image).
+type RegistrySnapshot struct {
+	Name   reg.RegistryName `yaml:"name,omitempty"`
+	Before []reg.Image      `yaml:"before,omitempty"`
```

Comment:
It doesn't seem like this field is used in this PR. Do we want to make sure that the populated registry matches the `Before` before we run the promotion tests. I'd also understand if you think that's overkill.

Reply:
Correct, we don't use `Before` in this PR. But thinking about this again, maybe it should. At the very least we would be verifying that the test-registries did get wiped clean (hermetic test environment). I'll work on this a bit.

Reply:
I just added a check to loop through all expected `Before` snapshots as well. It required a small change to how `getSnapshot` works (need to allocate 0-length array instead of nil array by default, because otherwise `reflect.DeepEqual` will fail), but is otherwise the same.

PTAL